File: iosApp/iosApp/Modules/NFTSend ✅/NFTSendWireframeFactory.swift
// Created by web3d4v on 04/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - NFTSendWireframeFactory

protocol NFTSendWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: NFTSendWireframeContext
    ) -> NFTSendWireframe
}

// MARK: - DefaultNFTSendWireframeFactory

final class DefaultNFTSendWireframeFactory {
    private let qrCodeScanWireframeFactory: QRCodeScanWireframeFactory
    private let confirmationWireframeFactory: ConfirmationWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService

    init(
        qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
        confirmationWireframeFactory: ConfirmationWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.qrCodeScanWireframeFactory = qrCodeScanWireframeFactory
        self.confirmationWireframeFactory = confirmationWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultNFTSendWireframeFactory: NFTSendWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: NFTSendWireframeContext
    ) -> NFTSendWireframe {
        DefaultNFTSendWireframe(
            parent,
            context: context,
            qrCodeScanWireframeFactory: qrCodeScanWireframeFactory,
            confirmationWireframeFactory: confirmationWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            networksService: networksService,
            currencyStoreService: currencyStoreService
        )
    }
}

// MARK: - Assember

final class NFTSendWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> NFTSendWireframeFactory in
            DefaultNFTSendWireframeFactory(
                qrCodeScanWireframeFactory: resolver.resolve(),
                confirmationWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                networksService: resolver.resolve(),
                currencyStoreService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/nftsend/NFTSendWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.nftsend

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.confirmation.ConfirmationWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframe
import com.sonsofcrypto.web3walletcore.modules.nftSend.NFTSendWireframeContext
import smartadapter.internal.extension.name

interface NFTSendWireframeFactory {
    fun make(parent: Fragment?, context: NFTSendWireframeContext): NFTSendWireframe
}

class DefaultNFTSendWireframeFactory(
    private val networksService: NetworksService,
    private val currencyStoreService: CurrencyStoreService,
    private val confirmationWireframeFactory: ConfirmationWireframeFactory,
): NFTSendWireframeFactory {

    override fun make(
        parent: Fragment?, context: NFTSendWireframeContext
    ): NFTSendWireframe = DefaultNFTSendWireframe(
        parent?.let { WeakRef(it) },
        context,
        networksService,
        currencyStoreService,
        confirmationWireframeFactory,
    )
}

class NFTSendWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {
        to.register(NFTSendWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultNFTSendWireframeFactory(
                it.resolve(NetworksService::class.name),
                it.resolve(CurrencyStoreService::class.name),
                it.resolve(ConfirmationWireframeFactory::class.name),
            )
        }
    }
}
